### PR TITLE
[ONNX] Use optional op to keep None in results for ONNX internal tests

### LINKED ISSUE #84130: [ONNX] ONNX output ignores None while Pytorch doesn't.
In ONNX, try to keep meaningless None node with Optional type as well. This can help to compare the results in internal test.

### PR DESCRIPTION
**Pull request by titaiwangms**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #84789
* #83186

All this time, PyTorch and ONNX has different strategy for None in output. And in internal test, we flatten the torch outputs to see if the rest of them matched. However, this doesn't work anymore in scripting after Optional node is introduced, since some of None would be kept.

#83184 forces script module to keep all Nones from Pytorch, but in ONNX, the model only keeps the ones generated with Optional node, and deletes those meaningless None.

This PR uses Optional node to keep those meaningless None in output as well, so when it comes to script module result comparison, Pytorch and ONNX should have the same amount of Nones.

cc @EikanWang @jgong5 @wenzhe-nrv @sanchitintel

### CODE REVIEW

File: torch/csrc/jit/passes/onnx/shape_type_inference.h [hunk 1]
```diff
@@ -57,7 +57,15 @@ TORCH_API void ONNXAssignOutputShape(
     at::ArrayRef<at::Tensor> outputs,
     const python::IODescriptor& desc,
     bool onnx_shape_inference,
-    bool is_script);
+    bool is_script,
+    int opset_version);
+
+// Replace None in output with Optional node (opset > 15) if it's
+// script model. This helps align the output format in internal tests.
```
Comment (kit1980) [inline torch/csrc/jit/passes/onnx/shape_type_inference.h:64]: Please clarify "internal tests". Which tests?

Comment (AllenTiTaiWang) [reply torch/csrc/jit/passes/onnx/shape_type_inference.h]: Done

**Comment by AllenTiTaiWang** [comment]
@BowenBao PTAL

File: test/onnx/test_pytorch_onnx_onnxruntime.py [hunk 2]
```diff
@@ -1599,10 +1599,10 @@ def forward(self, x: int, y: int):
...
         self.run_test(ArithmeticModule(), (x, y))
 
-    # Outputs that are always None are removed.
-    # Issue 84130: ONNX ignores mustNone() node, while pytorch
-    # doesn't, and that makes Optional comparison difficult to achieve.
-    @skipScriptTest()  # TODO Use onnx::Optional to replace erase None in shape_type_inference.cpp
+    # Trace: Outputs that are always None are removed.
+    # SciprtTest: Outputs that are always None are removed before opset 15.
+    # After opset 15, we replace the None in output with Optional node.
+    @skipScriptTest(15)
```
Comment (BowenBao) [inline test/onnx/test_pytorch_onnx_onnxruntime.py:1605]: nit: could utilize `reason` as an optional argument.

**Review by BowenBao — APPROVED** [review]

**Pull request closed without merging**